File: LBM_single_phase_beadspack.m
%% D2Q9 BGK LBM simulating single phase gravity driven flow through a 2D
%% cross section of a glass beads pack
%%
%% Author: Rui Xu, The University of Texas at Austin, 2017
%% This code is for your own use, please do not distribute.

%% clear the workspace
clear
close all
clc

tic

%% Initialization
%% Set parameters
tau = 1.0;               % relaxation time 
g = 0.00001;             % gravity or other force
density = 1;          
tf = 10001;            % maximum number of iteration steps
precision = 1e-5;      % covergence criterion
vold = 1000;
statIter = 100;        % check if convergence is achived every 100 time steps
data = imread('beads_dry.slice150.sub.tif'); % data is 0s and 255s

data = data/255; %get 0s and 1s from original data

% I will require fluid nodes to be 0, and 1 will be solid.
is_solid_node = data;    % define geometry

[ny,nx] = size(data);

% allocate memory for distribution function
f = zeros(ny,nx,9);   % particle distribution f1, f2, ... f9 for every point of the image. 
                      % for point data(i,j) related distributions are
                      % stored as f(i,j,1:9)
feq = f;              % equilibrium function feq
ftemp = feq;


% initialize distribution function
f(:,:,1) = (4./9. )*density;
f(:,:,2) = (1./9. )*density;
f(:,:,3) = (1./9. )*density;
f(:,:,4) = (1./9. )*density;
f(:,:,5) = (1./9. )*density;
f(:,:,6) = (1./36.)*density;
f(:,:,7) = (1./36.)*density;
f(:,:,8) = (1./36.)*density;
f(:,:,9) = (1./36.)*density;

% define lattice velocity vectors 
ex=zeros(9,1);
ey=zeros(9,1);
ex(0+1)= 0; ey(0+1)= 0;
ex(1+1)= 1; ey(1+1)= 0;
ex(2+1)= 0; ey(2+1)= 1;
ex(3+1)=-1; ey(3+1)= 0;
ex(4+1)= 0; ey(4+1)=-1;
ex(5+1)= 1; ey(5+1)= 1;
ex(6+1)=-1; ey(6+1)= 1;
ex(7+1)=-1; ey(7+1)=-1;
ex(8+1)= 1; ey(8+1)=-1;

%(each point has x component velocity ex
% and y componend ey
u_x = zeros(ny,nx);
u_y = zeros(ny,nx);

% node density
rho = zeros(ny,nx);

for ts=1:tf %Time loop
    
    ts
    
    % Computing macroscopic density, rho, and velocity, u=(ux,uy).
    for j=1:ny 
        
        for i=1:nx
            
            u_x(j,i) = 0.0;
            u_y(j,i) = 0.0;
            rho(j,i) = 0.0;
            
            if ~is_solid_node(j,i)  %fluid nodes
                
                for a=0:8
                    
                    rho(j,i) = rho(j,i) + f(j,i,a+1);
                                       
                    u_x(j,i) = u_x(j,i) + ex(a+1)*f(j,i,a+1);
                    u_y(j,i) = u_y(j,i) + ey(a+1)*f(j,i,a+1);
                    
                end
                
                u_x(j,i) = u_x(j,i)/rho(j,i);
                u_y(j,i) = u_y(j,i)/rho(j,i);
                
            end
                        
        end
    end
    
       
    % Compute the equilibrium distribution function, feq.
    f1=3.;
    f2=9./2.;
    f3=3./2.;
    
    for j=1:ny
        
        for i=1:nx
            
            if ~is_solid_node(j,i)
                
                rt0 = (4./9. )*rho(j,i);
                rt1 = (1./9. )*rho(j,i);
                rt2 = (1./36.)*rho(j,i);
                ueqxij =  u_x(j,i)+tau*g; %add forcing to velocity
                ueqyij =  u_y(j,i);
                uxsq   =  ueqxij * ueqxij;
                uysq   =  ueqyij * ueqyij;
                uxuy5  =  ueqxij +  ueqyij;
                uxuy6  = -ueqxij +  ueqyij;
                uxuy7  = -ueqxij + -ueqyij;
                uxuy8  =  ueqxij + -ueqyij;
                usq    =  uxsq + uysq;
                
                feq(j,i,0+1) = rt0*( 1.                              - f3*usq);
                feq(j,i,1+1) = rt1*( 1. + f1*ueqxij + f2*uxsq        - f3*usq);
                feq(j,i,2+1) = rt1*( 1. + f1*ueqyij + f2*uysq        - f3*usq);
                feq(j,i,3+1) = rt1*( 1. - f1*ueqxij + f2*uxsq         - f3*usq);
                feq(j,i,4+1) = rt1*( 1. - f1*ueqyij + f2*uysq         - f3*usq);
                feq(j,i,5+1) = rt2*( 1. + f1*uxuy5  + f2*uxuy5*uxuy5 - f3*usq);
                feq(j,i,6+1) = rt2*( 1. + f1*uxuy6  + f2*uxuy6*uxuy6 - f3*usq);
                feq(j,i,7+1) = rt2*( 1. + f1*uxuy7  + f2*uxuy7*uxuy7 - f3*usq);
                feq(j,i,8+1) = rt2*( 1. + f1*uxuy8  + f2*uxuy8*uxuy8 - f3*usq);
                
            end
        end
    end
        
    % Collision step.
    for j=1:ny
        
        for i=1:nx
            
            if is_solid_node(j,i) 
                
                % Standard bounceback
                
                temp   = f(j,i,1+1); f(j,i,1+1) = f(j,i,3+1); f(j,i,3+1) = temp;
                temp   = f(j,i,2+1); f(j,i,2+1) = f(j,i,4+1); f(j,i,4+1) = temp;
                temp   = f(j,i,5+1); f(j,i,5+1) = f(j,i,7+1); f(j,i,7+1) = temp;
                temp   = f(j,i,6+1); f(j,i,6+1) = f(j,i,8+1); f(j,i,8+1) = temp;
                
            else  
                % Regular collision away from solid boundary
                
                for a=1:9
                    
                    f(j,i,a) = f(j,i,a)-( f(j,i,a) - feq(j,i,a))/tau; 
                    
                end  
                
            end
        end          
    end
        
    % Streaming step; 
    for j=1:ny
        
        if j>1 
            jn = j-1;
        else
            jn = ny;
        end
        
        if j<ny
            jp = j+1;
        else 
            jp = 1;
        end
        
        for i=1:nx
                  
            if i>1
                in = i-1; 
            else 
                in = nx; 
            end
            if i<nx 
                ip = i+1; 
            else 
                ip = 1; 
            end
            
            
            ftemp(j,i,0+1)  = f(j,i,0+1);
            ftemp(j,ip,1+1) = f(j,i,1+1);
            ftemp(jp,i,2+1) = f(j,i,2+1);
            ftemp(j,in,3+1) = f(j,i,3+1);
            ftemp(jn,i ,4+1) = f(j,i,4+1);
            ftemp(jp,ip,5+1) = f(j,i,5+1);
            ftemp(jp,in,6+1) = f(j,i,6+1);
            ftemp(jn,in,7+1) = f(j,i,7+1);
            ftemp(jn,ip,8+1) = f(j,i,8+1);
            
        end
    end
    
    f=ftemp; 
  
    if rem(ts,statIter)==0
        u = sqrt(u_x.^2+u_y.^2);
        vnew = mean(mean(u));
        error = abs(vold-vnew)/vold;
        vold=vnew;
        
        imagesc(u)
        axis equal; 
        axis tight;
        colorbar;
        title('Speed profile');
        drawnow
        

        
        if error<precision
            disp('Simulation has converged');
            break;
        end
    end
    if ts == tf
        disp('Maximum iteration has reached');
    end
        
end

% plot geometry
figure
image(is_solid_node,'CDataMapping','scaled')
        axis tight
        axis equal
hold on;

% plot velocity profile
[x1,y1] = meshgrid(1:10:nx,1:10:ny);
quiver(x1,y1,u_x(1:10:ny,1:10:nx),u_y(1:10:ny,1:10:nx),'color','w','linewidth',1);
title('Velocity profile');

% plot streamlines
figure
image(is_solid_node,'CDataMapping','scaled')
        axis tight
        axis equal
hold on;
% Start a streamline seed every 10 pixels (feel free to change this
starty = 1:10:ny;
startx = ones(size(starty));
h = streamline(u_x,u_y,startx,starty);
set(h,'color','w','linewidth',1);
title('Streamlines');

toc

File: LBM_single_phase_beadspack.py
"""
D2Q9 BGK LBM simulation of single phase gravity driven flow through
a 2D cross section of a glass bead pack

Author: Bernard Chang, The University of Texas at Austin
Adopted from: Rui Xu, The University of Texas at Austin
"""

# Import packages
import numpy as np
# import jax as np
import matplotlib.pyplot as plt
from plotting_utils import plot_profile, plot_quiver, plot_streamlines
import scipy as sc
import tifffile
from time import perf_counter_ns
import tqdm


def run_lbm(data):
    
    # Initialization
    tau = 1.0  # Relaxation time
    g = 0.00001  # Gravity or other force
    density = 1.
    tf = 10001  # Maximum number of iteration steps
    precision = 1.E-5  # Convergence criterion
    vold = 1000
    eps = 1E-6
    
    check_convergence = 30  # Check convergence every [check_convergence] time steps
    
    
    # Indices of fluid nodes
    fluid_id = np.argwhere(data == 0)
    fx = fluid_id[:, 0]
    fy = fluid_id[:, 1]
    
    # Indices of solid nodes
    solid_id = np.argwhere(data == 1)
    sx = solid_id[:, 0]
    sy = solid_id[:, 1]
    
    # Solid nodes are labeled 1, fluid nodes are labeled 0
    is_solid_node = data
    
    nx, ny = data.shape
    
    # Initialize distribution functions
    f = np.array([4./9.,
                  1./9.,
                  1./9.,
                  1./9.,
                  1./9.,
                  1./36.,
                  1./36.,
                  1./36.,
                  1./36.], dtype=np.double) * density
    
    # Broadcast to 3D array with each slice corresponding to
    f = np.tile(f[:, np.newaxis, np.newaxis], (nx, ny)).astype(np.single)
    
    # Allocate memory to equilibrium functions
    feq = np.empty_like(f, dtype=np.double)
    
    # Define lattice velocity vectors. Transpose to make it a column vector
    ex = np.array([0, 1, 0, -1, 0, 1, -1, -1, 1], dtype=np.double)
    ey = np.array([0, 0, 1, 0, -1, 1, 1, -1, -1], dtype=np.double)
    
    # Each point has x-component ex, and y-component ey
    u_x = np.empty((nx, ny), dtype=np.double)
    u_y = np.empty((nx, ny), dtype=np.double)
    
    # Node density
    rho = np.zeros((nx, ny), dtype=np.double)
    
    # For equilibrium distribution function (feq) calculation 
    f1 = 3.
    f2 = 9./2.
    f3 = 3./2.
    
    # # Begin time loop
    tic = perf_counter_ns()
    for ts in tqdm(range(tf)):
        # print(f"{ts = }")  # Print timestep
    
        # Compute macroscopic density, rho and velocity.
        u_x = np.zeros((nx, ny), dtype=np.double)
        u_y = np.zeros((nx, ny), dtype=np.double)
        rho = np.zeros((nx, ny), dtype=np.double)
        
        rho[fx, fy] += np.sum(f[:, fx, fy], axis=0)
        u_x[fx, fy] += np.sum(ex[:, None] * f[:, fx, fy], axis=0)
        u_y[fx, fy] += np.sum(ey[:, None] * f[:, fx, fy], axis=0)
        
        u_x[fx, fy] = u_x[fx, fy] / rho[fx, fy]
        u_y[fx, fy] = u_y[fx, fy] / rho[fx, fy]
        
    
    
        # Add forcing to velocity
        ueqxij = u_x[fx, fy] + tau * g
        ueqyij = u_y[fx, fy]
        uxsq = ueqxij ** 2
        uysq = ueqyij ** 2
        uxuy5 = ueqxij + ueqyij
        uxuy6 = -ueqxij + ueqyij
        uxuy7 = -ueqxij + (-ueqyij)
        uxuy8 = ueqxij + (-ueqyij)
        usq = uxsq + uysq
        
        # Compute equilibrium distribution function, feq
        rt0 = 4./9. * rho[fx, fy]
        rt1 = 1./9. * rho[fx, fy]
        rt2 = 1./36. * rho[fx, fy]
        
        feq[0, fx, fy] = rt0*(1. - f3 * usq)
        feq[1, fx, fy] = rt1*(1. + f1 * ueqxij + f2 * uxsq - f3 * usq)
        feq[2, fx, fy] = rt1*(1. + f1 * ueqyij + f2 * uysq - f3 * usq)
        feq[3, fx, fy] = rt1*(1. - f1 * ueqxij + f2 * uxsq - f3 * usq)
        feq[4, fx, fy] = rt1*(1. - f1 * ueqyij + f2 * uysq - f3 * usq)
        feq[5, fx, fy] = rt2*(1. + f1 * uxuy5  + f2 * uxuy5 * uxuy5 - f3 * usq)
        feq[6, fx, fy] = rt2*(1. + f1 * uxuy6  + f2 * uxuy6 * uxuy6 - f3 * usq)
        feq[7, fx, fy] = rt2*(1. + f1 * uxuy7  + f2 * uxuy7 * uxuy7 - f3 * usq)
        feq[8, fx, fy] = rt2*(1. + f1 * uxuy8  + f2 * uxuy8 * uxuy8 - f3 * usq)
    
        # Collision Step
        # Standard Bounceback for Solid Nodes
        # Left-Right
        tmp = f[1, sx, sy]
        f[1, sx, sy] = f[3, sx, sy]
        f[3, sx, sy] = tmp
        
        # Up-Down
        tmp = f[2, sx, sy]
        f[2, sx, sy] = f[4, sx, sy]
        f[4, sx, sy] =  tmp
        
        # Top Right - Bottom Left
        tmp = f[5, sx, sy]
        f[5, sx, sy] = f[7, sx, sy]
        f[7, sx, sy] =  tmp
        
        # Top Left - Bottom Right
        tmp = f[6, sx, sy]
        f[6, sx, sy] = f[8, sx, sy]
        f[8, sx, sy] =  tmp
        
        # Regular collision in fluid nodes
        f[:, fx, fy] -= (f[:, fx, fy] - feq[:, fx, fy]) / tau
    
    
        # Streaming Step
        f[1] = np.roll(f[1], 1, axis=1)
        f[2] = np.roll(f[2], 1, axis=0)
        f[3] = np.roll(f[3], -1, axis=1)
        f[4] = np.roll(f[4], -1, axis=0)
        
        f[5] = np.roll(f[5], (1, 1), axis=(0,1))
        f[6] = np.roll(f[6], (-1, 1), axis=(1,0))
        f[7] = np.roll(f[7], (-1, -1), axis=(0,1))
        f[8] = np.roll(f[8], (1, -1), axis=(1,0))
        
        u = np.sqrt(u_x**2 + u_y**2)
        # Plot the time step and check convergence every check_convergence time step
        if ts % check_convergence == 0:
            
            vnew = np.mean(u)
            error = np.abs(vold - vnew) / (vold+eps)
            vold = vnew
    
            if error < precision:
                print(f'Simulation has converged in {ts} time steps')
                break
    
        if ts == tf:
            print('Reached maximum iterations')
        
    toc = perf_counter_ns()
    print(f"Elapsed Time: {(toc - tic)*1E-9}s")
    
    return u_x, u_y, u


if __name__ == '__main__':

    # Read in and create geometry
    data = tifffile.imread("beads_dry.slice150.sub.tif")  # Data labels are 0 and 255
    data = data / 255  # Make Data 0 and 1s
    # data = data.astype(np.uint8)
    # data = data[0:100, :]
    u_x, u_y, u = run_lbm(data)
    # profile_fig = plot_profile(u, cmap='jet')
    # quiver_fig = plot_quiver(u_x, u_y)
    streamline_fig = plot_streamlines(u_x, u_y, arrowsize=0, density=6)
    
    # profile_fig.savefig("output_figures/velocity_profile.png")
    # quiver_fig.savefig("output_figures/velocity_field.png")
    streamline_fig.savefig("output_figures/streamlines.png")
    
    # plt.figure(dpi=300)
    # plt.imshow(u, cmap='jet')
    # plt.tight_layout()
    # plt.colorbar()
    # plt.title('Speed Profile')
    # plt.show()


